### PR TITLE
Added flexible options for the occupancy ratio calculation for nomina…

### PR DESCRIPTION
**Pull request by efranzis**
…tor and denominator. Fix also indentation

@fkrizek please have a look if this works for you.
We can change now the options with 
SetAreaCalculationDetails
SetAreaCalculationDetails(0, 0) is what you want, this is the default

### CODE REVIEW

**Comment by efranzis** [comment]
Update:

Comment taken from PR https://github.com/alisw/AliPhysics/pull/8055
Discussion about that should continue here

Dear Markus (@mfasDa) and Filip (@fkrizek),

just to summarize the issue what was done before and how things are proposed to change in the proper rho task:

**For the Area calculation of physical jets:**
Before
See: https://github.com/alisw/AliPhysics/blob/db4782541921fd4e9590848316679ae11ec28d8b/PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx
1.	Loop over jets
2.	Exclude leading jets, if selected
3.	Sum the area of all physical jets 
4.	Exclude jets not suited for the analysis
5.	Exclude background jets overlapping with signal jets from the analysis

Now
https://github.com/alisw/AliPhysics/blob/df6bb15ee1083b277d6d5c39019a67b4a34c1c14/PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx
If fExcludeAreaExcludedJets==0
1.	Loop over jets
2.	Add area of all physical jets
3.	Exclude jets not suited for the analysis
4.	Exclude leading jets, if selected
5.	Exclude background jets overlapping with signal jets from the analysis

If fExcludeAreaExcludedJets==1
1.	Loop over jets
2.	Exclude jets not suited for the analysis
3.	Exclude leading jets, if selected
4.	Exclude background jets overlapping with signal jets from the analysis
5.	Add area of all remaining physical jets


**For the Area calculation of all jets as a proxy for the occupied area:**
Before
See: https://github.com/alisw/AliPhysics/blob/db4782541921fd4e9590848316679ae11ec28d8b/PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx
1.	Loop over jets
2.	Exclude leading jets, if selected
3.	Sum the area of all jets physical+ghost
4.	Exclude jets not suited for the analysis
5.	Exclude background jets overlapping with signal jets from the analysis

Now:
https://github.com/alisw/AliPhysics/blob/df6bb15ee1083b277d6d5c39019a67b4a34c1c14/PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx
1.	Loop over jets
2.	Add area of all jets (physical+ghost)
3.	Exclude jets not suited for the analysis
4.	Exclude leading jets, if selected
5.	Exclude background jets overlapping with signal jets from the analysis


The old approach is inconsistent to include some jets and exclude others. 
Since you seem to agree with the treatment of the physical area and we only need to discuss the area of all jets (physical+ghosts) I would suggest to included the total area physical+ghost  in the fExcludeAreaExcludedJets Condition. Then we would have the same options as above for the physical area.
And then you can run your analysis with fExcludeAreaExcludedJets==1.

**My suggestion:**
A.	@fkrizek  Please look at option fExcludeAreaExcludedJets
B.	If you still want some third way of how things are done as indicated in your e-mail yesterday (exclude some jets include others), please directly comment on the proper pull request so that we have things documented and can move forward
C.	As a reminder the Link to the proper pull request:
https://github.com/alisw/AliPhysics/pull/8012
D.	I will add this comment here in the original pull request that is still waiting to be merged and we should continue discussing there and close this pull request here.

Regards,
Eliane

File: PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.h [hunk 1]
```diff
@@ -66,9 +66,10 @@ class AliAnalysisTaskRhoSparse : public AliAnalysisTaskRhoBase {
   		);
 
   void             UserCreateOutputObjects();
-  void             SetExcludeLeadJets(UInt_t n)         { fNExclLeadJets = n       ; }
-  void             SetExcludeOverlapJets(Bool_t input)  { fExcludeOverlaps = input ; }
-  void             SetRhoCMS(Bool_t cms)                { fRhoCMS = cms            ; }
+  void             SetExcludeLeadJets(UInt_t n)           { fNExclLeadJets = n       ; }
+  void             SetExcludeOverlapJets(Bool_t input)    { fExcludeOverlaps = input ; }
+  void             SetRhoCMS(Bool_t cms)                  { fRhoCMS = cms            ; }
+  void             SetAreaCalculationDetails(Bool_t inputA, Bool_t inputB){ fUseTPCArea = inputA ; fExcludeAreaExcludedJets = inputB; }
```
Comment (mfasDa) [inline PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.h]: Please use self-explaining parameter names. inputA and inputB are not easy to digest.

File: PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx [hunk 2]
```diff
@@ -22,22 +22,26 @@ ClassImp(AliAnalysisTaskRhoSparse)
...
-  fExcludeOverlaps(0),
-  fRhoCMS(0),
-  fHistOccCorrvsCent(0)
+AliAnalysisTaskRhoBase("AliAnalysisTaskRhoSparse"),
+fNExclLeadJets(0),
+fExcludeOverlaps(0),
+fRhoCMS(0),
+fUseTPCArea(0),
+fExcludeAreaExcludedJets(0),
+fHistOccCorrvsCent(0)
```
Comment (mfasDa) [inline PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx]: Why is this formatting applied? Code is better readable with spaces for the attribute definition.

File: PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx [hunk 3]
```diff
@@ -22,22 +22,26 @@ ClassImp(AliAnalysisTaskRhoSparse)
...
-  fExcludeOverlaps(0),
-  fRhoCMS(0),
-  fHistOccCorrvsCent(0)
+AliAnalysisTaskRhoBase(name, histo),
+fNExclLeadJets(0),
+fExcludeOverlaps(0),
+fRhoCMS(0),
+fUseTPCArea(0),
+fExcludeAreaExcludedJets(0),
+fHistOccCorrvsCent(0)
```
Comment (mfasDa) [inline PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx]: Same here, unnecessary reformatting.

File: PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx [hunk 4]
```diff
@@ -131,57 +134,88 @@ Bool_t AliAnalysisTaskRhoSparse::Run()
   static Double_t rhovec[999];
```
Comment (mfasDa) [inline PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx]: Same comment as in other pull request - the large static array should be replaced by std::vector as it is excessive resource usage and particularly generates problems in parallelized environments or when multiple objects of the same type are used in parallel.

**Comment by fkrizek** [comment]
Dear Elian,
I think  that if I exclude something as a signal jet,   then  neither area of this jet   nor   momentum of this jet should contribute to rho evaluation.   So in priciple rho should be evaluated just based on information from the remaining part of the event. 

In line 150    you are addind together area of all jets,   including area of those that you are going to exclude later on.     
Would it be possible to apply the L161+L162   befor L150?

F.

**Comment by efranzis** [comment]
Hi @fkrizek,

so just to double check instead of doing
(Option1)
1. Loop over jets
2. Exclude leading jets, if selected
3. Exclude jets not suited for the analysis
4. Exclude background jets overlapping with signal jets from the analysis
5. Add area of all jets (physical+ghost)
as I suggested above (by including it in the fExcludeAreaExcludedJets condition)

You want to do:
(Option2 Your suggestion)
1. Loop over jets
2. Exclude leading jets, if selected
3. Add area of all jets (physical+ghost)
4. Exclude jets not suited for the analysis
5. Exclude background jets overlapping with signal jets from the analysis

The lines you wanted to change (_apply the L161+L162 befor L150_) as you write might not be what you actually want to do. 
Please double check my comment here and select the option (option1 or option2) you want.
Or let me know if you want to deviate from the two options I gave you.


Can

**Comment by fkrizek** [comment]
Hi @efranzis,

option 1   what i want.    However looking to the code 
https://github.com/alisw/AliPhysics/blob/df6bb15ee1083b277d6d5c39019a67b4a34c1c14/PWGJE/EMCALJetTasks/AliAnalysisTaskRhoSparse.cxx
i see that    summing of jet area  (total area + physics jet) is happening before  exclusion of 
the leading jets.     Am i looking to the wrong code version?


Cheers Filip

**Comment by efranzis** [comment]
Hi @fkrizek ,

yes currently it is summed before things are excluded. Only the physical jet area hast the option to exclude before or after the summation.

What I meant in my long comment above is

> The old approach is inconsistent to include some jets and exclude others.
> Since you seem to agree with the treatment of the physical area and we only need to discuss the area of all jets (physical+ghosts) I would suggest to included the total area physical+ghost in the fExcludeAreaExcludedJets Condition. Then we would have the same options as above for the physical area.
> And then you can run your analysis with fExcludeAreaExcludedJets==1.

So I will then go ahead and include Markus's comment and will add the total jet area also in the
condition of the fExcludeAreaExcludedJets checks.

I just want to say that our new approach will be a bit different than the original task
(which I think is good) because the old task did inconsistently include some jets and exclude others for the total area calculation. 

See

> For the Area calculation of all jets as a proxy for the occupied area:
> Before ...

In my previous long comment.

Kind Regards,
Eliane

**Comment by mfasDa** [comment]
Hi @fkrizek,

Please let me know if the PR is fine with you and can be merged! If you want changes in the code please use the review interface (go to "Files changed", select line you want to change and click on the "+" symbol) - this makes the discussion easier.

Cheers

Markus

**Comment by efranzis** [comment]
Hi @mfasDa,

I don't know why my additional changes now did not automatically merge into the current PR but they instead opened a new one #8062
If I close this PR does the conversation here get lost?

**Comment by mfasDa** [comment]
Hi @efranzis,

you use different branches for the two pull requests. A pull request however gets only updated if the branch is the same. My recommendation: Open a new branch before doing a pull request and delete the branch once the pull request is closed - this allows you to update the pull request whenever desired. 

Please do the following: Close the pull reques 8062, merge the branch of that pull request (ppJet) into the one for this pull request (master) and push the master branch. You can also do this on github opening a pull request for efranzis:ppJet -> efranzis:master, then this pull request gets also updated.

Cheers

Markus

**Comment by mfasDa** [comment]
+1

**Pull request merged**